core/src/main/scala/io/projectglow/vcf/VCFStreamWriter.scala

```diff
@@ -66,25 +64,74 @@ class VCFStreamWriter(
     .build
 
   def write(vc: VariantContext): Unit = {
-    if (!headerHasBeenSetOrWritten) {
-      maybeSetHeaderSampleIds(vc)
+    val vcBuilder = new VariantContextBuilder(vc)
+    val iterator = vc.getAttributes.entrySet().iterator()
+    while (iterator.hasNext) { // parse to string, then write,
+      // otherwise the write messes up double precisions
+      val entry = iterator.next()
+      vcBuilder.attribute(
+        entry.getKey,
+        VariantContextToVCFRowConverter.parseObjectAsString(entry.getValue))
+    }
+
+    if (header == null) {
+      val (sampleIds, missing) = if (!inferSampleIds) {
+        sampleIdsMissingOpt.get
+      } else {
+        val vcSamples = vcBuilder.getGenotypes.asScala.map(_.getSampleName)
+        val numPresentSamples = vcSamples.count(!_.isEmpty)
+        val numAllSamples = vcSamples.size
+        numPresentSamples match {
+          case 0 => (VCFWriterUtils.getMissingSampleIds(numAllSamples), true)
+          case `numAllSamples` => (vcSamples.sorted, false)
+          case _ =>
+            throw new IllegalArgumentException("Cannot mix missing and non-missing sample IDs.")
+        }
+      }
+      headerSampleSet = new JHashSet(sampleIds.asJava)
+      header = new VCFHeader(headerLineSet.asJava, sampleIds.asJava)
+      replaceSampleIds = missing
       if (writeHeader) {
         writer.writeHeader(header)
       } else {
         writer.setHeader(header)
       }
-      headerHasBeenSetOrWritten = true
     }
 
-    val maybeReplacedVC = maybeReplaceDefaultSampleIds(vc)
-    val vcBuilder = new VariantContextBuilder(maybeReplacedVC)
-    val iterator = maybeReplacedVC.getAttributes.entrySet().iterator()
-    while (iterator.hasNext) { // parse to string, then write,
-      // otherwise the write messes up double precisions
-      val entry = iterator.next()
-      vcBuilder.attribute(
-        entry.getKey,
-        VariantContextToVCFRowConverter.parseObjectAsString(entry.getValue))
+    if (replaceSampleIds) {
```

Comment:
Also, I think that this could all be simplified by splitting into two steps:
1. If we're replacing sample IDs, build a new genotype array.
2. If we've inferred sample IDs, iterate through the maybe rebuilt genotype array and verify that the sample ids match what we inferred.

Reply:
I believe that if 1 is true, then we don't need 2.